### PR TITLE
Remove redundant exception object creation

### PR DESCRIPTION
**Pull request by GrandFisher**
`List<NoShardAvailableActionException> ` was just use its `size()`.It seems that there is no need to create such a list, just use an int to count directly to achieve the same effect

### CODE REVIEW

**Comment by GrandFisher** [comment]
@elasticsearchmachine

**Comment by idegtiarenko** [comment]
@elasticmachine ok to test

File: server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java [hunk 1]
```diff
@@ -267,7 +267,7 @@ protected void doExecute(Task task, Request request, ActionListener<Response> li
         private final Map<String, List<ShardRouting>> nodeIds;
         private final AtomicReferenceArray<Object> responses;
         private final AtomicInteger counter = new AtomicInteger();
-        private List<NoShardAvailableActionException> unavailableShardExceptions = new ArrayList<>();
+        private int unavailableShardCount = 0;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:270]: @GrandFisher do you mind making `unavailableShardCount` final to be consistent with the other fields?
Unavailable shards could be counted with a local variable and then assigned to this field in the end of the constructor.

Comment (GrandFisher) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:270]: @idegtiarenko  Thanks for your review, I'll fix this

Comment (GrandFisher) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java]: @idegtiarenko  I have fixed this and please review again. By the way, how can I improve the code to be able to pass the test, can you give me some advice ?

Comment (GrandFisher) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java]: May I rebase the master and git push force ?

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java]: Please merge the latest master into your branch.
This should trigger another build

Comment (GrandFisher) [reply server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java]: @idegtiarenko  It seems pass all code checks, thanks for your advice.

**Comment by GrandFisher** [comment]
@idegtiarenko  Hi,  I have finished the optimization and merge the newest master code.
Please help to review if you have free time. Thanks.

**Review by idegtiarenko — APPROVED** [review]
@GrandFisher thanks for Your contribution! I will merge this pr shortly.

**Comment by DaveCTurner** [comment]
I'm backporting this to 7.17 because it seems safe enough to do so and will resolve conflicts with other backports.

**Pull request merged**